File: app-android/src/main/java/com/ustadmobile/port/android/view/SalePaymentRecyclerAdapter.java
package com.ustadmobile.port.android.view;

import android.arch.paging.PagedListAdapter;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.AppCompatImageButton;
import android.support.v7.widget.AppCompatImageView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.app.Activity;
import android.widget.PopupMenu;
import android.widget.TextView;

import com.toughra.ustadmobile.R;
import com.ustadmobile.core.controller.SaleDetailPresenter;

import com.ustadmobile.core.util.UMCalendarUtil;
import com.ustadmobile.lib.db.entities.SalePayment;

public class SalePaymentRecyclerAdapter extends
        PagedListAdapter<SalePayment,
                SalePaymentRecyclerAdapter.SaleDetailViewHolder> {

    Context theContext;
    Activity theActivity;
    SaleDetailPresenter mPresenter;

    @NonNull
    @Override
    public SaleDetailViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {


        View list = LayoutInflater.from(theContext).inflate(
                R.layout.item_key_value, parent, false);
        return new SaleDetailViewHolder(list);

    }

    @Override
    public void onBindViewHolder(@NonNull SaleDetailViewHolder holder, int position) {

        SalePayment entity = getItem(position);

        String prettyDate = UMCalendarUtil.getPrettyDateSuperSimpleFromLong(
                        entity.getSalePaymentPaidDate(), null);
        TextView keyTV = holder.itemView.findViewById(R.id.item_key_value_key);
        TextView valueTV = holder.itemView.findViewById(R.id.item_key_value_value);

        String amountText = entity.getSalePaymentPaidAmount() + " " +
                entity.getSalePaymentCurrency();
        keyTV.setText(prettyDate);
        valueTV.setText(amountText);

        AppCompatImageView dots = holder.itemView.findViewById(R.id.item_key_value_context_dots);

        //Options to Edit/Delete every schedule in the list
        dots.setOnClickListener((View v) -> {
            //creating a popup menu
            PopupMenu popup = new PopupMenu(theActivity.getApplicationContext(), v);
            popup.setOnMenuItemClickListener(item -> {
                int i = item.getItemId();
                if (i == R.id.edit) {
                    mPresenter.handleEditPayment(entity.getSalePaymentUid());
                    return true;
                } else if (i == R.id.delete) {
                    mPresenter.handleDeletePayment(entity.getSalePaymentUid());
                    return true;
                } else {
                    return false;
                }
            });
            //inflating menu from xml resource
            popup.inflate(R.menu.menu_edit_delete);

            popup.getMenu().findItem(R.id.edit).setVisible(true);

            //displaying the popup
            popup.show();
        });
    }

    protected class SaleDetailViewHolder extends RecyclerView.ViewHolder {
        protected SaleDetailViewHolder(View itemView) {
            super(itemView);
        }
    }

    protected SalePaymentRecyclerAdapter(
            @NonNull DiffUtil.ItemCallback<SalePayment> diffCallback,
            SaleDetailPresenter thePresenter,
            Activity activity,
            Context context) {
        super(diffCallback);
        mPresenter = thePresenter;
        theContext = context;
        theActivity = activity;
    }


}

File: app-android/src/main/java/com/ustadmobile/port/android/view/SalePaymentRecyclerAdapter.kt

package com.ustadmobile.port.android.view


import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.toughra.ustadmobile.databinding.ItemSalepaymentBinding
import com.ustadmobile.core.controller.SalePaymentItemListener
import com.ustadmobile.lib.db.entities.SalePaymentWithSaleItems
import com.ustadmobile.port.android.view.binding.MODE_START_OF_DAY


class SalePaymentRecyclerAdapter(var itemListener: SalePaymentItemListener?)
    : ListAdapter<SalePaymentWithSaleItems,
        SalePaymentRecyclerAdapter.InventoryTransactionDetailHolder>(DIFF_CALLBACK) {

    class InventoryTransactionDetailHolder(val itemBinding: ItemSalepaymentBinding)
        : RecyclerView.ViewHolder(itemBinding.root)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): InventoryTransactionDetailHolder {
        val itemBinding = ItemSalepaymentBinding.inflate(
                LayoutInflater.from(parent.context), parent, false)
        itemBinding.listener = itemListener
        itemBinding.dateTimeMode = MODE_START_OF_DAY
        itemBinding.timeZoneId = "Asia/Kabul"
        return InventoryTransactionDetailHolder(itemBinding)
    }

    override fun onBindViewHolder(holder: InventoryTransactionDetailHolder, position: Int) {
        val item = getItem(position)
        holder.itemBinding.salePayment = item
        holder.itemBinding.dateTimeMode = MODE_START_OF_DAY
        holder.itemBinding.timeZoneId = "Asia/Kabul"
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        super.onDetachedFromRecyclerView(recyclerView)
        itemListener = null
    }

    companion object {
        val DIFF_CALLBACK: DiffUtil.ItemCallback<SalePaymentWithSaleItems> = object
            : DiffUtil.ItemCallback<SalePaymentWithSaleItems>() {
            override fun areItemsTheSame(oldItem: SalePaymentWithSaleItems,
                                         newItem: SalePaymentWithSaleItems): Boolean {
                return oldItem.salePaymentUid == newItem.salePaymentUid

            }

            override fun areContentsTheSame(oldItem: SalePaymentWithSaleItems,
                                            newItem: SalePaymentWithSaleItems): Boolean {
                return oldItem == newItem
            }
        }
    }

}